react-blog-app/src/hooks/useArticle.tsx

```diff
@@ -0,0 +1,51 @@
+import { useState, useEffect, useRef } from "react";
+import axios from "axios";
+
+const useArticle = (id:any) => {
+
+    const baseURL = `${process.env.REACT_APP_API_URL}/api/posts/${id}`;
+
+    const isCurrent = useRef(true);
+    const [article, setArticle]:any = useState({});
+    const [isLoading, setIsLoading] = useState(false);
+    const [error, setError] = useState(null);
+
+    useEffect(() => {
+        const prevTitle = document.title;
+        //preventing title change to undefined
+        document.title = article.postTitle ?? prevTitle;
+        return () => {
+            document.title = prevTitle;
+        };
+    }, [article]);
+
+    useEffect(() => {
+        return () => {
+            isCurrent.current = false;
+        };
+    }, []);
+
+    useEffect(() => {
+        const fetchArticle = async () => {
+            setIsLoading(true);
+
+            try {
+                const { data } = await axios.get(baseURL);
+                //Preventing changing state of an unmounted component
+                if (isCurrent.current) {
+                    setArticle(data);
+                    setIsLoading(false);
+                }
```

Comment:
Something's wrong with this hack. First of all, it looks like a hack 😃 
Is it something about the react warning message?
Usually, I don't pay attention to this warning. Many projects with modals/popups have that warning, it doesn't break anything, so usually, I don't try to fix it.
And if I would try to fix it, I will probably start with changing the data flow.

But I found the solution how you can fix it that looks less hacky
https://stackoverflow.com/a/60907638

although I wouldn't do anything with that warning 😄

Reply:
It wasn't actually related to the "memory leak" warning and popups. I was just trying to learn more about different hooks and to find use cases for them in my project. I have read that useRef hook is also used to prevent state change on an unmounted component. So basically that's what i did here.

For example, let's image that the connection is really slow: a user opens a page that starts fetching data, and before the data fetching is complete—the user goes to another page. As a result, when finally the data is fetched—our component will try to update it's state and will fail, because it was previously unmounted.

I just did not know that it isn't a good practise to do so 😬.

For example, one of the devs on react team suggests a [similar solution](https://github.com/facebook/react/issues/14369#issuecomment-468267798), which is a bit simpler and did not utilize useRef, but useRef is also mentioned there as a possible solution in this comment thread.

Reply:
And I don’t really quite understand why is this a hack in the first place? Is it because of it’s unnecessity and because it is not usually done that way on commercial projects?

And btw I don't mean to sound contradictory, I just want to figure it out.

Reply:
> Is it because of it’s unnecessity and because it is not usually done that way on commercial projects?

Yes, usually nobody pays attention to that warning.
And I never saw workarounds for that warning

In that thread, there is a [comment](https://github.com/facebook/react/issues/14369#issuecomment-458868231) with the solution I consider the only solution which is not a workaround or a hack, it's like I said, changing the data flow.

Why it also looks hacky to me – if imaging the situation you described, with a slow connection and moving through pages, we need to add this useRef logic to every request we have in our project. And that looks like the wrong abstraction, probably we need to handle that behavior in the different place, for example, closer to the actual request. Especially axios already has aborting request functionality.

Reply:
Thank you so much for the explanation 😄! I have thought about the consequent need to reuse it in every component too, and it surely felt wrong. But I just couldn't understand why this solution should or shouldn't be used. I also did not know about axios' aborting request functionality.